enhancements/baremetal/baremetal-keepalived-unicast.md

```diff
@@ -0,0 +1,193 @@
+2. The keepalived-monitor container should be updated to [retrieve](https://github.com/openshift/baremetal-runtimecfg/blob/8242fd84d6c6d267d2d5551abf8e4f778b4f0615/pkg/config/node.go#L219-#L244) the nodes (both compute and control) IP addresses from the kube-apiserver and render them to Keepalived config.
+
+### Risks and Mitigations
+
+It is important to not destabilize our existing installation method, that uses
+multicast successfully. Full parity is expected before landing.
+
+Since upgrade may trigger a Keepalived configuration migration process (see upgrade section below), it's important to add upgrade coverage for bare metal IPI in CI.
+
+## Design Details
+
+### Test Plan
+
+Because of the fundamental inclusion of this in Baremetal IPI installations,
+any breakage by CI before landing can probably be assumed to be caused by
+this work.
+
+Take down nodes both in active as in passive mode and observe the VIP move correctly and without significantly increasing the current expected downtime in VIP changes.
+
+Force a disaster recovery scenario and see that we can recover the cluster following the usual procedure.
+
+Since the baremetal-runtimecfg is used by all on-prem platforms, we should verify that Ovirt, vSphere, and OpenStack still successfully deployed in Keepalived multicast mode.
+
+### Upgrade / Downgrade Strategy
+
+In case that the Keepalived config file was already generated by the previous SW version, Keepalived will first continue to run in a mode determined in accordance with the existing file.
+If this mode is different from the mode specified in MCO templates a Keepalived configuration migration process should be triggered.
+
+The next steps illustrate Keepalived flow in case of upgrade/downgrade:
+1. If config file exists determined_mode=mode determined from the file else determined_mode=mode specified in MCO templates
+2. Run Keepalived in determined_mode.
+3. If determined_mode != desired_mode (mode specified in MCO templates) initiate a configuration migration process.
+
+#### Configuration migration process
+
+Since multicast and unicast are considered as separate Keepalived domains and will not interoperate, the updated configuration should simultaneously reload in all the relevant Keepalived instances. This option is feasible since the nodes run NTP.
+
+The next steps describe how synced reload should be implemented:
+
+1. One of the nodes will take the role of coordinator (the node with the lowest IP address).
+2. The coordinator should be responsible for:
+    - Deciding if simultaneously reload is required.
+    - Verifying that all nodes are ready for reload.
+    - Choosing the timing for reload and updating the rest of the nodes.
```

Comment:
How will updating the rest of the nodes be done?

Reply:
A new TCP server will be added to the keepalived-monitor container, this server will serve as a kind of control channel.
The coordinator node will use this server for communicating with rest of the nodes.

Reply:
I'm not too comfortable with the idea of a new custom bespoke TCP service just for this use case.  We should always hesitate before adding new network services.

Have you considered a more kubernetes native approach to this problem?  I imagine pods all looking at a shared configmap that indicates what mode and configuration they should be running in, something like that?

In other words, what alternatives have been considered here?

Reply:
@russellb 

Well, currently we don't use any CRD nor configmap for configuring and running keepalived (Keepalived should run early in bootstrap phase to provide access to temp control plane), we use static pod capability and configuration is rendered by the MCO. 

An alternative method could be to use the MachineConfig CRD for triggering the migration, 
create MachineConfig CR, this CR will create a new file (includes the desired mode + planned reload time) in all nodes and keepalived-monitor will do the rest of the work.

What do you think about this option?

Reply:
Yes, that sounds better.  I'd just like to see something built around existing interfaces

Reply:
The original idea was to just have a MachineConfig that writes a config file that the keepalived monitor reads. Once the monitor reads it and notices that the requested mode differs from the configured one, it checks if the other nodes are in the same Node object `CurrentConfig`. Once all the monitors see that the other nodes are in the same currentConfig, they should be able to perform the reload.

This solution just uses the OCP API, i.e., a MachineConfig and the Node annotation for MC rendered currentconfig and desiredconfig.

This means that the user would create a MachineConfig and once all the machines have rebooted and have the new config, the change can take place. Personally I'd prefer if it was optional to set a time for when to do the reload, as it may involve a bit of guesswork of knowing when the MachineConfig rollout will end (to know the earliest time).